lib/ukallocbbuddy/bbuddy.c

```diff
@@ -260,6 +278,10 @@ static void *bbuddy_palloc(struct uk_alloc *a, unsigned long num_pages)
 	b->free_head[i] = alloc_ch->next;
 	alloc_ch->next->pprev = alloc_ch->pprev;
 
+	map_alloc(b, (uintptr_t)alloc_ch, 1UL << order);
+
+	uk_spin_unlock(&b->level_locks[i]);
```

Comment:
Could mention in a comment where the corresponding "uk_spin_lock()" for this "uk_spin_unlock()" comes from e.g. the for loop above. Or can unlock earlier (e.g. before "break") and add another set of lock() / unlock() for this codebase block, to keep lock / unlock pairs.

The "map_alloc()" call was moved earlier in the codebase flow to be part of this synchronized code block? Does this reorder have any effects given the follow-up while loop?

Reply:
I don't think unlocking before the break and then locking again would be safe. In this time window, another thread could enter the `if (!FREELIST_EMPTY(b->free_head[i]))` statement and find the same free block, resulting in a race condition.

Yes, the `map_alloc()` call was moved so that it is part of the synchronized  code block. Otherwise, if the free block that has been selected for this allocation resulted from a previous split, there will be a chance that it will be merged by a `free()` running on a parallel thread, even if we are in the process of allocating it (see [here](https://github.com/unikraft/unikraft/pull/801/files#diff-7a854c8df048714d0efcab86bedab15bf77f36afe24270e946fd1bdb578c14afR355)).

The reorder shouldn't have any effects. The loop only works on lower orders of the allocator, while the `map_alloc` call makes sure that this block won't be merged with its buddy and added to a higher order list.

Reply:
Thank you, Razvan, for the explanation and the updates.